File: amlip_cpp/src/cpp/types/status/StatusDataType.cpp
// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*!
 * @file StatusDataType.cpp
 */

#include <types/status/StatusDataType.hpp>

#include <fastdds/rtps/common/CdrSerialization.hpp>

#include <cpp_utils/utils.hpp>

namespace eprosima {
namespace amlip {
namespace types {

const char* StatusDataType::TYPE_NAME_ = "status";

StatusDataType::StatusDataType(
        const AmlipIdDataType& id,
        const NodeKind& node_kind,
        const StateKind& state)
    : id_(id)
    , node_kind_(node_kind)
    , state_(state)
{
}

StatusDataType::StatusDataType()
    : StatusDataType(
        AmlipIdDataType(),
        NodeKind::undetermined,
        StateKind::unknown)
{
}

bool StatusDataType::operator ==(
        const StatusDataType& x) const
{
    return (id() == x.id() && node_kind() == x.node_kind() && state() == x.state());
}

bool StatusDataType::operator !=(
        const StatusDataType& x) const
{
    return !(*this == x);
}

AmlipIdDataType StatusDataType::id() const noexcept
{
    return id_;
}

AmlipIdDataType& StatusDataType::id() noexcept
{
    return id_;
}

NodeKind StatusDataType::node_kind() const noexcept
{
    return node_kind_;
}

NodeKind& StatusDataType::node_kind() noexcept
{
    return node_kind_;
}

StateKind StatusDataType::state() const noexcept
{
    return state_;
}

StateKind& StatusDataType::state() noexcept
{
    return state_;
}

bool StatusDataType::is_defined() const noexcept
{
    return (id().is_defined() && node_kind() != NodeKind::undetermined && state() != StateKind::unknown);
}

std::string StatusDataType::to_string() const noexcept
{
    return utils::generic_to_string(*this);
}

bool StatusDataType::is_bounded()
{
    return true;
}

bool StatusDataType::is_plain()
{
    return true;
}

bool StatusDataType::construct_sample(
        void* memory)
{
    return new (memory) StatusDataType();
}

std::string StatusDataType::type_name()
{
    return TYPE_NAME_;
}

std::ostream& operator <<(
        std::ostream& os,
        const StatusDataType& st)
{
    os << "STATUS{" << st.id() << ", " << st.node_kind() << ", " << st.state() << "}";
    return os;
}

} /* namespace types */
} /* namespace amlip */
} /* namespace eprosima */

// Include auxiliary functions like for serializing/deserializing.
#include  <types/status/impl/StatusDataTypeCdrAux.ipp>

File: amlip_py/amlip_py/types/StatusDataType.py
# Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""AML-IP Status data type API specification."""

from enum import Enum

from amlip_py.types.AmlipIdDataType import AmlipIdDataType

from amlip_swig import NodeKind_agent, NodeKind_computing, NodeKind_discovery, NodeKind_main
from amlip_swig import NodeKind_meta, NodeKind_status, NodeKind_undetermined
from amlip_swig import StateKind_dropped, StateKind_running, StateKind_stopped, StateKind_unknown
from amlip_swig import StatusDataType as cpp_StatusDataType


class NodeKind(Enum):
    """Enumeration to represent the different Node Kinds of each Node."""

    UNDETERMINED = NodeKind_undetermined
    DISCOVERY = NodeKind_discovery
    AGENT = NodeKind_agent
    MAIN = NodeKind_main
    COMPUTATIONAL = NodeKind_computing
    EDGE = NodeKind_status
    STATUS = NodeKind_meta


class StateKind(Enum):
    """Enumeration to represent the different State of each Node."""

    UNKNOWN = StateKind_unknown
    RUNNING = StateKind_running
    STOPPED = StateKind_stopped
    DROPPED = StateKind_dropped


class StatusDataType(cpp_StatusDataType):
    """
    Object that represents the Status of a node.

    A Status is formed by:
    - Node kind
    - Node State
    - Node Id
    """

    def __init__(
            self,
            id: AmlipIdDataType = None,
            node_kind: NodeKind = None,
            state: StateKind = None):
        """
        Construct a new Status object.

        By default it is created a non defined object.
        If callee sets the 3 arguments, it creates a defined State with such values

        Parameters
        ----------
        id: AmlipIdDataType
            Id of the new Status
            [Default = None]

        node_kind: NodeKind = None
            Node Kind of the new Status
            [Default = None]

        state: StateKind = None)
            State of the new Status
            [Default = None]
        """
        if not id or not node_kind or not state:
            super().__init__()
        else:
            super().__init__(id, node_kind, state)

    def get_id(self) -> AmlipIdDataType:
        """Get Id associated with this Status."""
        return cpp_StatusDataType.id(self)

    def get_node_kind(self) -> NodeKind:
        """Get Node Kind associated with this Status."""
        return cpp_StatusDataType.node_kind(self)

    def get_status_kind(self) -> StateKind:
        """Get Status Kind associated with this Status."""
        return cpp_StatusDataType.status_kind(self)
